### ISSUE TITLE
Is their anyway to disable the auto-complete/suggestions or have customised prompts?

### ISSUE THREAD

**Issue by ItsPugle** [fish-shell/fish-shell#4473]
Basically just the title lol

**Comment by zanchey** [comment]
Disabling the autosuggestion is tracked in #1363; to create a custom prompt, create a `fish_prompt` function - see the [FAQ](https://fishshell.com/docs/current/faq.html#faq-prompt).

**Comment by floam** [comment]
You can customize your prompt in `fish_config`. We do not provide a means for disabling autosuggestions. Is there something about them that is annoying you? Our preference would be to try to make them better rather than providing an on/off switch.

**Comment by ItsPugle** [comment]
@zanchey > I don't want to create a custom prompt, I want to get rid of them, fullstop.

@floam > I've had a look at `fish_config`, but it still isn't going for me :( I just don't like the autocomplete/suggestions because it clogs up my terminal and is quite annoying when I'm trying to navigate to a specific directory with subdirectories.

**Comment by floam** [comment]
Do you mean that fish_config does not work? The "just a dollar" prompt is very short with just a "$" symbol. Or do you mean none suffice for you?

If you want a totally empty prompt, just make a function that does nothing.

```shell
function fish_prompt; end;
```

**Comment by floam** [comment]
Assuming your terminal is black, if you simply do not want to notice the autosuggestions, just change the color:

```fish
set fish_color_autosuggestion black
```

**Comment by ItsPugle** [comment]
No, no - not at all; `fish_config` works perfectly fine. The prompts are good, I just don't need suggestions as well. I'll try the set fish thing like you suggested, only white.

**Comment by faho** [comment]
The question here has been answered.

**Comment by vkalinsky** [comment]
@ItsPugle did you figure out how to do it? This question is closed as "answered", but I don't see an appropriate answer here.

**Comment by ItsPugle** [comment]
Nope @vkalinsky! I just ended up ditching fish

**Comment by joeyfernau** [comment]
I'd like to disable it because my terminal emulator does not fare well with it; changing the color to black will still emit the characters to the terminal emulator. It is not a solution in my case (even though it is a solution for people with better terminals).

**Comment by floam** [comment]
What terminal are you using?

**Comment by joeyfernau** [comment]
@floam, I was using an older version of fish, version 2.2.0. See #5331.

**Comment by faho** [comment]
@dogweather: For recording screen demos, wouldn't switching history work?

**Comment by dogweather** [comment]
@faho Thanks - but I'm not sure what you mean, and couldn't find "switching history" with a Google search. Do you mean moving a history file of some kind?

**Comment by dogweather** [comment]
Thanks @zanchey but is that in an upcoming version?

```bash
$ fish --private
fish: unrecognized option '--private'
$ fish --version
fish, version 2.7.1
```

**Comment by floam** [comment]
3.0, which is released.

**Comment by zanchey** [comment]
It's in the recently-released 3.0.0.

Without it, you can set `fish_history` to change history sessions, or use `set -g fish_history ''` to disable history saving completely.

**Comment by boxabirds** [comment]
How do I get fish to do this:

ln -s /app /Users/administrator/dev/thing

When I type "/app" it forces it to be "/Applications". I can't seem to override this.

**Comment by faho** [comment]
@boxabirds: This is not the correct issue for that question, but anyway:

What exactly are you doing? If you just type `ln -s /app<SPACE>`, fish should not be changing anything - we don't expand abbreviations except for in command position (and they can't contain spaces). If that expands anything, you've installed a third-party thing that does it.

If you type `ln -s /app<TAB>`, then that's tab-completion, which is case-insensitive, so it's doing the correct thing. If you want it to be `/app`, then type `/app` and don't press tab.

If you type `ln -s /app<RIGHT-ARROW>`, then that's accepting the suggestion, which should look like `ln -s /Applications...`. So don't do that.

TL;DR: Just type the thing.

**Comment by vkalinsky** [comment]
@faho the problem is when you type, say, `ln -s /app`, it shows `ln -s /Applications`. So it's not just "type the thing", it's "type the thing and don't trust what's in your terminal".

**Comment by faho** [comment]
>it shows ln -s /Applications. So it's not just "type the thing", it's "type the thing and don't trust what's in your terminal".

Yes, it shows it greyed-out. That's the "suggestion", and it's not really part of the commandline. If you press enter, what will be executed is the part *before* that.

**Comment by boxabirds** [comment]
Thanks for a quick response! Yes I was confused by the autosuggestion. It's
super cool but I really got stuck. Here's one thought: maybe experiment
with a tiny delay so the suggestion doesn't appear until maybe 0.2s later?

Never mind there's no delay in Gmail's suggestion so maybe it's just a
learning curve thing...

**Comment by vkalinsky** [comment]
@faho but it shows it as a part of command line, and it's confusing.

**Comment by faho** [comment]
@vkalinsky: It doesn't confuse most people, and even then it just takes a minute to get used to it (barring #3134).

Anyway, we've decided we're not going to offer an option to disable it, so further discussion isn't helpful.

**Comment by afontenot** [comment]
Since https://github.com/fish-shell/fish-shell/issues/1363 was locked I'm posting this here instead. Hopefully anyone in that thread who is interested will be able to find this.

I have found a workaround for this issue that is probably the most effective thing short of patching fish and recompiling it. I'm posting it here in the hope that it's useful to someone. Just stick the following snippet I wrote in your `config.fish`

```
# I say we take off and nuke the entire site from orbit.
# It's the only way to be sure.
set -l cp1252_chars ' !"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQR'\
    'STUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~€‚ƒ„…†‡ˆ‰Š‹ŒŽ‘’“”•–—˜™š›'\
    'œžŸ¡¢£¤¥¦§¨©ª«¬®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâ'\
    'ãäåæçèéêëìíîïðñòóôõö÷øùúûüýþ'
for c in (string split '' $cp1252_chars)
    bind $c self-insert suppress-autosuggestion
end
```

It's very ugly but also rather clever. We rebind every single normal key in fish to insert itself and then suppress the autosuggestions. I'm honestly shocked that this works as well as it does. It not only doesn't cause any problems with using the rebound keys (that I have noticed), but tab completions continue to work and ctrl-r still searches the history. Despite setting a lot of binds it does not noticeably slow down fish (and only applies to interactive mode anyway).

For this example I chose to use the characters from the Windows-1252 character set, since they cover pretty much everything that users with western languages would ever expect to enter at a terminal prompt. I generated the list with a Python one-liner. You could of course modify the list to contain the characters that you use regularly.

**Comment by vp2177** [comment]
I couldn't find it in this thread, but `fish` has a setting to turn autosuggestions off:

```fish
set fish_autosuggestion_enabled 0
```

You're welcome 🐠

**Comment by afontenot** [comment]
@vp2177 that's terrific, thanks!

Looks like it was quietly added in [3.4.0](https://github.com/fish-shell/fish-shell/releases/tag/3.4.0) but none of the old reports requesting the feature were marked fixed.
